### ISSUE TITLE
cdc_acm module -> No USB Device

### ISSUE THREAD

**Issue by h4waii** [AndProx/AndProx#8]
I'm unable to get AndProx to recognize that the PM3 is plugged in. I know it works, as I can reach it through Termux with a natively compiled proxmark3 client, without issue, on /dev/ttyACM0. 

`lsusb` shows the correct IDs `9ac4:4b8f` that your client supports.

Any idea why it wouldn't detect correctly? Related to cdc_adm module?

This is with a Redmi Note 4 running Lineage-14.1 with a custom kernel that DOES support USB Host Mode. I've also tested and got it working on a Pixel, so I know the firmware, cabling, etc, is not the issue.

**Comment by micolous** [comment]
You're correct, the `cdc_acm` module is the cause of this issue. `usb-serial-for-android` [implements a userspace CDC ACM driver](https://github.com/mik3y/usb-serial-for-android), rather than using kernel facilities.

What'll be happening is that Android's USB stack won't be able to take control of the device, so it will never even get presented to a userspace Android application in that fashion.

Does it work if you `rmmod cdc_acm`?

**Comment by h4waii** [comment]
I strangely can't rmmod it "Function not implemented, and `modprobe` doesn't exist.

**Comment by h4waii** [comment]
I strangely can't rmmod it "Function not implemented, and `modprobe` doesn't exist.

**Comment by h4waii** [comment]
I strangely can't rmmod it "Function not implemented, and `modprobe` doesn't exist.

**Comment by bogiton** [comment]
I seem to be having the same issue with the exact same phone.
The problem is that I can't rmmod the cdc_acm module, since my device is not rooted. Is there any other solution we could try?

**Comment by micolous** [comment]
@bogiton What device and Android build, version and distribution (eg: vendor supplied, Lineage OS) are you using that has `cdc_acm` built by default?

**Comment by bogiton** [comment]
My device is the Xiaomi Redmi Note 4, running stock MIUI 9.5 (Android version 6.0). The thing is that I tried a sample app using the USB Host API and it worked fine, but only after explicitly requesting the permissions for the connected device. I also noticed that AndProx is asking for permission the moment I disconnect the device, which of course is not persistent when I re-connect it. When I find time I may debug it to see what the actual problem is with AndProx, but in the meantime I'm open to any suggestion. :)

**Comment by h4waii** [comment]
While I haven't been able to unbind cdc_adm properly, would it be possible for you to add a root request in order to interface directly with the port? Possibly have an option to request root and specify a port in the Settings menu?

**Comment by bogiton** [comment]
Sadly, not possible. My device is not rooted. TBH, that was the whole point of AndProx :P

**Comment by micolous** [comment]
@bogiton I think you have a different issue, which may be addressed by https://github.com/Proxmark/proxmark3/pull/565. There was mention that this issue impacted some Huawei devices, and the "permission prompt on disconnect" is similar to what I saw with a Sony device that is impacted by this issue.

The firmware which I attached to AndProx 2.0.2 _does **not** have this patch_, because it predates it by a few months. If you checkout https://github.com/Proxmark/proxmark3/tree/f3dc6d59f45e25fd199cb20cdf4d3af270d93963, that's what I last tested and should have minimal changes.

You can verify whether you have that patch by attaching to a PC, and [if you see `Product: proxmark3`, then you have the patch](https://github.com/Proxmark/proxmark3/pull/565#issuecomment-366430026).  If you see `Product: PM3`, then you **don't** have the patch.

@h4waii Yeah, that is what I was proposing.

At the moment PM3 folks are working on integrating something similar to my other changes, so that there's a consistent API with which to override the serial device and start up the PM3 client.

Next is to get AndProx working on that API (see branch `pwpiwi-comms-refactor` for the work-in-progress version based on proposed PRs), then I can start working from PM3 `master` once again, and start working on more drastic changes to AndProx internals, like getting the cdc_acm module working.

**Comment by bogiton** [comment]
@micolous thanks for the suggestion, but I am using the iceman fork, where the product description is set as "PM3 Device". I suppose that only the master proxmark3 repo is supported. When I get a chance to flash that one, I will let you know if it worked.

**Comment by h4waii** [comment]
Just want to update, I've moved to LineageOS 15.1, which is based on Android 8.1 Oreo, and no longer have this issue. I can connect the device and access it just fine with AndProx now.

**Comment by micolous** [comment]
@h4waii Thanks for confirming. I'm going to track getting this working with cdc_acm as another issue, #13.

@bogiton Always use PM3 with the same firmware and client version, as there is no protocol spec! Iceman's firmware has some incompatible PM3 commands which can cause undefined behaviour.

Getting iceman's fork working is tracked in #12, and making people use the right firmware is tracked in #14.
